File: C++/46 Permutations.cpp
class Solution
{
public:
    vector<vector<int>> permute(vector<int> &nums)
    {
        int n = nums.size();
        vector<vector<int>> ans;
        vector<int> path;
        vector<bool> flags(n, false);
        auto dfs = [&](this auto &&dfs, int cnt) -> void
        {
            if (cnt == nums.size())
            {
                ans.push_back(path);
                return;
            }

            for (int i = 0; i < n; i++)
            {
                if (!flags[i])
                {
                    flags[i] = true;
                    path.push_back(nums[i]);
                    dfs(cnt + 1);
                    path.pop_back();
                    flags[i] = false;
                }
            }
        };
        dfs(0);
        return ans;
    }
};

File: Swift/46 Permutations.swift
class Solution {
    func permute(_ nums: [Int]) -> [[Int]] {
        var ans: [[Int]] = .init()
        var path: [Int] = .init()
        let n = nums.count
        var st: [Bool] = .init(repeating: false, count: n)
        func dfs(_ cnt: Int) {
            if cnt == n {
                ans.append(path)
                return
            }

            for (i, _) in nums.enumerated() {
                if !st[i] {
                    st[i] = true
                    path.append(nums[i])
                    dfs(cnt + 1)
                    path.popLast()
                    st[i] = false
                }
            }
        }
        dfs(0)
        return ans
    }
}
